### PR TITLE
feat: Fix cost calculations for AWS S3

### PR DESCRIPTION
**Pull request by bishal7679**
## Problem #1022

Currently cost for AWS S3 is always showing as $0.
 
## Solution

This PR is fixing that issue by adding more filters and usage metrics for request to s3 bucket with absolute pricemap fetching.

## Changes Made

- Fetching price info using AWS SDK price module using `AmazonS3` service code and appropriate filters.
- getting price map using util function GetPriceMap from [utils.go](https://github.com/tailwarden/komiser/blob/develop/providers/aws/utils/utils.go)
- fetching resource objects using S3 client
- also fetching usage metrics as request using cloudwatch
- Finally added calculation as per all the metrics

## Checklist

- [x] Code follows the <a href="https://github.com/tailwarden/komiser/blob/master/CONTRIBUTING.md">contributing</a> guidelines
- [x] Changes have been thoroughly tested
- [ ] <a href="https://github.com/tailwarden/docs.komiser.io">Documentation</a> has been updated, if necessary
- [ ] Any dependencies have been added to the project, if necessary

Taking ref from :- #1053

### CODE REVIEW

File: providers/aws/s3/buckets.go [hunk 1]
```diff
@@ -9,69 +9,138 @@ import (
...
 		if sizeInTB <= 50 {
-			monthlyCost = (sizeInTB * 1000) * 0.023
+			storageCostPerGB = (sizeInTB * 1024) * 0.023
 		} else if sizeInTB <= 450 {
-			monthlyCost = (sizeInTB * 1000) * 0.022
+			storageCostPerGB = (sizeInTB * 1024) * 0.022
 		} else {
-			monthlyCost = (sizeInTB * 1000) * 0.021
+			storageCostPerGB = (sizeInTB * 1024) * 0.021
+		}
```
Comment (Azanul) [inline providers/aws/s3/buckets.go:105]: make it dynamic using the pricing output

File: providers/aws/s3/buckets.go [hunk 2]
```diff
@@ -9,69 +9,138 @@ import (
...
+			Namespace:  aws.String("AWS/S3"),
+			Dimensions: []cloudwatchTypes.Dimension{
+				{
+					Name:  aws.String("BucketName"),
+					Value: bucket.Name,
+				},
+				{
+					Name:  aws.String("StorageType"),
+					Value: aws.String("StandardStorage"),
+				},
```
Comment (Azanul) [inline providers/aws/s3/buckets.go:122]: Storage type might be different

File: providers/aws/s3/buckets.go [hunk 3]
```diff
@@ -9,69 +9,138 @@ import (
...
+			Dimensions: []cloudwatchTypes.Dimension{
+				{
 					Name:  aws.String("BucketName"),
 					Value: bucket.Name,
 				},
-				types.Dimension{
+				{
 					Name:  aws.String("StorageType"),
 					Value: aws.String("StandardStorage"),
 				},
```
Comment (Azanul) [inline providers/aws/s3/buckets.go:80]: Storage type might be different

File: providers/aws/s3/buckets.go [hunk 4]
```diff
@@ -9,69 +9,138 @@ import (
...
+			MetricName: aws.String("AllRequests"),
+			Namespace:  aws.String("AWS/S3"),
+			Dimensions: []cloudwatchTypes.Dimension{
+				{
+					Name:  aws.String("BucketName"),
+					Value: bucket.Name,
+				},
+				{
+					Name:  aws.String("StorageType"),
+					Value: aws.String("StandardStorage"),
```
Comment (AvineshTripathi) [inline providers/aws/s3/buckets.go:121]: Are we making this static?

**Review by Azanul — CHANGES_REQUESTED** [review]
https://docs.aws.amazon.com/AmazonS3/latest/userguide/metrics-dimensions.html#s3-cloudwatch-metrics

File: providers/aws/s3/buckets.go [hunk 5]
```diff
@@ -9,69 +9,117 @@ import (
...
+		requestCount := 0.0
+		if metricsUsageOutput != nil && len(metricsUsageOutput.Datapoints) > 0 {
+			requestCount = *metricsUsageOutput.Datapoints[0].Sum
+		}
+		// requestCost := (requestCount / 1000) * 0.0004
+
+		monthlyCost := 0.0
+
+		storageCharges := awsUtils.GetCost(priceMap["AWS-S3-Storage"], sizeInTB*1024)      // charges per GB
+		requestCharges := awsUtils.GetCost(priceMap["AWS-S3-Requests"], requestCount/1000) // charges per 1000 request
```
[the remark this answers is not in the document]
Comment (bishal7679) [reply providers/aws/s3/buckets.go:121]: should I use "group" as a key for both or only its for requests @Azanul

[the remark this answers is not in the document]
Comment (Azanul) [reply providers/aws/s3/buckets.go:121]: >  The group from product attributes is used as key for the pricing map

check the linked pricingOutput attached.

[the remark this answers is not in the document]
Comment (bishal7679) [reply providers/aws/s3/buckets.go:121]: yes I checked it. I can see group attribute. But for which one storage or request I should use "group" key as its only associated with requests which I can see

Comment (Azanul) [inline providers/aws/s3/buckets.go:121]:
```suggestion
		requestCharges := awsUtils.GetCost(priceMap["S3-API-Tier1"], requestCount)
```

Comment (bishal7679) [reply providers/aws/s3/buckets.go]: isn't Tier1 being static??

Comment (Azanul) [reply providers/aws/s3/buckets.go]: It is indeed, I think a bucket can store data in different storage tiers so to fetch full data with correct pricing we'll have to iterate through all storage classes to calculate the correct cost. I apologize for any confusion that I might have caused.
The storage classes can be found here https://docs.aws.amazon.com/AmazonS3/latest/userguide/metrics-dimensions.html#s3-cloudwatch-metrics.
I'll also have to go through the pricingOutput myself to find the correct corresponding keys for the priceMap

Comment (Azanul) [reply providers/aws/s3/buckets.go]: I think we need to scope the issue appropriately. It won't be perfect after this merge but it'll definitely be a step in the right direction.

[the remark this answers is not in the document]
Comment (Azanul) [reply providers/aws/s3/buckets.go]: That's something that you need to figure out. I do not have a sure shot way of doing this either other than looking throught the pricingOutput. If you find out a proper way or documentation do let me know.

**Review by Azanul — APPROVED** [review]
I think the code comments are pretty obvious and can be removed. Thanks for the work @bishal7679

**Review by mlabouardy — APPROVED** [review]
Thanks for the changes @bishal7679 once the comments are removed, I'll merge the PR :)

**Comment by bishal7679** [comment]
All good to go now :+1: @mlabouardy

**Pull request merged**